File: java/src/main/java/qiniu/uip/db/format/ipdb/Reader.java
package qiniu.uip.db.format.ipdb;

import qiniu.uip.db.IPFormatException;
import qiniu.uip.db.InvalidDatabaseException;
import qiniu.uip.db.IpInfo;
import qiniu.uip.db.NotFoundException;
import qiniu.uip.util.Json;

import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.HashMap;


final class Reader {

    private static final int cacheDepth = 12;
    MetaData meta;
    private int fileSize;
    private int nodeCount;
    private byte[] data;
    private int v4offset;
    private int[] ipv4BitsCache;

    private HashMap<Integer, IpInfo> resultCache;

    Reader(byte[] data) throws InvalidDatabaseException {
        this.data = data;
        this.fileSize = data.length;
        if (this.fileSize < 5) {
            throw new InvalidDatabaseException("database file size error");
        }

        long metaLength = bytesToLong(
                this.data[0],
                this.data[1],
                this.data[2],
                this.data[3]
        );

        try {
            int end = Long.valueOf(metaLength).intValue() + 4;
            byte[] metaBytes = Arrays.copyOfRange(this.data, 4, end);

            MetaData meta = Json.parseObject(new String(metaBytes), MetaData.class);

            this.nodeCount = meta.node_count;
            this.meta = meta;
        } catch (Exception e) {
            throw new InvalidDatabaseException(e.getMessage());
        }

        if ((meta.total_size + Long.valueOf(metaLength).intValue() + 4) != this.data.length) {
            throw new InvalidDatabaseException("database file size error");
        }

        this.data = Arrays.copyOfRange(this.data, Long.valueOf(metaLength).intValue() + 4, this.fileSize);

        /** for ipv4 */
        if (0x01 == (this.meta.ip_version & 0x01)) {
            int node = 0;
            for (int i = 0; i < 96 && node < this.nodeCount; i++) {
                if (i >= 80) {
                    node = this.readNode(node, 1);
                } else {
                    node = this.readNode(node, 0);
                }
            }
            this.v4offset = node;
            initCache();
        } else {
            this.ipv4BitsCache = null;
            this.v4offset = 0;
        }
    }

    private static byte[] indexToBytes(int i) {
        int i1 = (i << (16 - cacheDepth)) >> 8;
        int i2 = 0xFF & (i << (16 - cacheDepth));
        return new byte[]{(byte) i1, (byte) i2};
    }

    private static long bytesToLong(byte a, byte b, byte c, byte d) {
        return int2long((((a & 0xff) << 24) | ((b & 0xff) << 16) | ((c & 0xff) << 8) | (d & 0xff)));
    }

    private static long int2long(int i) {
        long l = i & 0x7fffffffL;
        if (i < 0) {
            l |= 0x080000000L;
        }
        return l;
    }

    void buildCache(byte[][] ipList) {
        this.resultCache = new HashMap<>();
        for (byte[] ip : ipList) {
            try {
                query(ip, false, true);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    IpInfo query(byte[] ip, boolean cache, boolean build) throws IPFormatException, InvalidDatabaseException {
        int node = find0(ip);
        if (cache && this.resultCache != null) {
            IpInfo info = this.resultCache.get(node);
            if (info != null) {
                return info;
            }
        }
        String[] parts = resolveNode(node, "CN");
        if (parts == null) {
            throw new IPFormatException("invalid ip address");
        }
        String[] fields = getSupportFields();
        IpInfo info = new IpInfo();
        for (int i = 0, l = parts.length; i < l; i++) {
            switch (fields[i]) {
                case "country_name":
                    info.country = parts[i];
                    break;
                case "region_name":
                    info.province = parts[i];
                    break;
                case "city_name":
                    info.city = parts[i];
                    break;
                case "isp_domain":
                    info.isp = parts[i];
                    break;
                case "asn":
                    info.asn = parts[i];
                case "line":
                    info.line = parts[i];
                    break;
                case "district":
                    info.district = parts[i];
                    break;
                case "continent_code":
                    switch (parts[i]) {
                        case "AS":
                            info.continent = "亚洲";
                            break;
                        case "EU":
                            info.continent = "欧洲";
                            break;
                        case "NA":
                            info.continent = "北美洲";
                            break;
                        case "SA":
                            info.continent = "南美洲";
                            break;
                        case "AF":
                            info.continent = "非洲";
                            break;
                        case "OC":
                            info.continent = "大洋洲";
                            break;
                        case "AN":
                            info.continent = "南极洲";
                            break;
                        default:
                            info.continent = parts[i];
                    }
                    break;
            }
        }
        if (build && this.resultCache != null) {
            this.resultCache.put(node, info);
        }
        return info;
    }

    private int bytesToIndex(byte[] b) {
        int i1 = (0xFF & (int) (b[0])) << 8 >> (16 - cacheDepth);
        int i2 = (0xFF & (int) (b[1])) >> (16 - cacheDepth);
        return i1 | i2;
    }

    private void initCache() {
        ipv4BitsCache = new int[1 << cacheDepth];
        //construct cache from binary trie tree for reduce read memory time
        for (int i = 0; i < ipv4BitsCache.length; i++) {
            byte[] b = indexToBytes(i);
            int node = readDepth(v4offset, cacheDepth, 0, b);
            ipv4BitsCache[i] = node;
        }
    }

    int find0(byte[] ip) throws IPFormatException, InvalidDatabaseException {
        if (ip.length == 16) {
            if (!isIPv6()) {
                throw new IPFormatException("no support ipv6");
            }
        } else if (ip.length == 4) {
            if (!isIPv4()) {
                throw new IPFormatException("no support ipv4");
            }
        } else {
            throw new IPFormatException("ip format error");
        }

        // find node
        int node = 0;
        try {
            node = this.findNode(ip);
        } catch (NotFoundException nfe) {
            return 0;
        }

        return node;
    }

    String[] find(byte[] addr, String language) throws IPFormatException, InvalidDatabaseException {
        int node = find0(addr);

        return resolveNode(node, language);
    }

    private String[] resolveNode(int node, String language) throws InvalidDatabaseException {
        int off;
        try {
            off = this.meta.languages.get(language);
        } catch (NullPointerException e) {
            return null;
        }

        final int resolved = node - this.nodeCount + this.nodeCount * 8;
        if (resolved >= this.fileSize) {
            throw new InvalidDatabaseException("database resolve error");
        }

        byte b = 0;
        int size = Long.valueOf(bytesToLong(
                b,
                b,
                this.data[resolved],
                this.data[resolved + 1]
        )).intValue();

        if (this.data.length < (resolved + 2 + size)) {
            throw new InvalidDatabaseException("database resolve error");
        }

        final String data = new String(this.data, resolved + 2, size, StandardCharsets.UTF_8);
        return Arrays.copyOfRange(data.split("\t", this.meta.fields.length * this.meta.languages.size()), off, off + this.meta.fields.length);
    }

    private int readDepth(int node, int depth, int i, byte[] binary) {
        for (; i < depth; i++) {
            if (node >= this.nodeCount) {
                break;
            }
            node = this.readNode(node, 1 & ((0xFF & binary[i / 8]) >> 7 - (i % 8)));
        }
        return node;
    }

    private int findNode(byte[] binary) throws NotFoundException {

        int node = 0;

        final int bit = binary.length * 8;

        if (bit == 32) {
            node = this.v4offset;
        }
        int i = 0;
        if (ipv4BitsCache != null && bit == 32) {
            int index = bytesToIndex(binary);
            node = ipv4BitsCache[index];
            if (node > this.nodeCount) {
                return node;
            }
            i = cacheDepth;
        }
        node = readDepth(node, bit, i, binary);
        if (node > this.nodeCount) {
            return node;
        }

        throw new NotFoundException("ip not found");
    }

    private int readNode(int node, int index) {
        int off = node * 8 + index * 4;

        return Long.valueOf(bytesToLong(
                this.data[off],
                this.data[off + 1],
                this.data[off + 2],
                this.data[off + 3]
        )).intValue();
    }

    boolean isIPv4() {
        return (this.meta.ip_version & 0x01) == 0x01;
    }

    boolean isIPv6() {
        return (this.meta.ip_version & 0x02) == 0x02;
    }

    int getBuildUTCTime() {
        return this.meta.build;
    }

    String[] getSupportFields() {
        return this.meta.fields;
    }

    String getSupportLanguages() {
        return this.meta.languages.keySet().toString();
    }
}

File: rust/src/formats/ipdb/reader.rs
use std::collections::HashMap;
use std::io;
use std::io::{Result, Error, ErrorKind};

use byteorder::{ByteOrder, BigEndian};

use serde::{Deserialize};

use crate::data::{IpInfo};

#[derive(Deserialize)]
pub struct MetaData {
    #[serde(skip_deserializing)]
    pub meta_size: usize,
    pub build: i64,
    pub ip_version: i32,
    pub node_count: i32,
    pub languages: HashMap<String, i32>,
    pub fields: Vec<String>,
    pub total_size: i32,
    pub version: Option<String>,
    pub extra: Option<Vec<String>>,
}

impl MetaData {
    fn new(buf: &[u8]) -> Result<MetaData> {
        if buf.len() < 4 {
            return Err(Error::new(ErrorKind::InvalidData, "meta data length < 4"));
        }

        let meta_size = BigEndian::read_i32(&buf[0..4]) as usize;
        let end = meta_size + 4;
        if buf.len() < end {
            return Err(Error::new(ErrorKind::InvalidData, "meta data length < meta_size"));
        }
        let v:serde_json::Result<MetaData> = serde_json::from_slice(&buf[4..end]);
        return match v {
            Ok(mut m) => {
                m.meta_size = meta_size;
                Ok(m)
            },
            Err(e) => {
                Err(Error::new(ErrorKind::InvalidData, e))
            }
        }
    }
}

const CACHE_DEPTH: i32 = 12;

pub struct Reader {
    pub meta: MetaData,
    file_size: i32,
    node_count: i32,
    data: Vec<u8>,
    v4_offset: i32,
    ipv4_bits_cache: Vec<i32>,
    _result_cache: HashMap<i32, IpInfo>,
}

impl Reader{
    pub fn new(bytes: &[u8]) -> Result<Reader> {
        let m = MetaData::new(&bytes)?;
        if m.total_size + m.meta_size as i32 +4 != bytes.len() as i32 {
            return Err(Error::new(ErrorKind::InvalidData, "invalid data length"));
        }
        let data = (&bytes[m.meta_size + 4..]).to_vec();

        let mut reader = Reader {
            file_size: bytes.len() as i32,
            node_count: m.node_count,
            v4_offset: 0,
            ipv4_bits_cache: Vec::with_capacity(1<<CACHE_DEPTH),
            _result_cache: HashMap::new(),
            data,
            meta: m,
        };
        if reader.is_ipv4() {
            let mut node = 0;
            for i in 0..96 {
                if i >= 80 {
                    node = reader.read_node(node, 1);
                } else {
                    node = reader.read_node(node, 0);
                }
                if node >= reader.node_count {
                    break;
                }
            }
            reader.v4_offset = node;
            reader.init_cache()
        } else {
            reader.v4_offset = 0
        }
        Ok(reader)
    }

    fn read_node(&self, node: i32, index:i32) -> i32 {
        let off = node * 8 + index * 4;
        BigEndian::read_i32(&self.data[off as usize..off as usize + 4])
    }

    fn init_cache(&mut self) {
        //construct cache from binary trie tree for reduce read memory time
        let mut b: [u8;2] = [0,0];
        for i in 0..(1<<CACHE_DEPTH) {
            // index to bytes
            b[0] = ((i << (16 - CACHE_DEPTH)) >> 8) as u8;
            b[1] = (0xFF & (i << (16 - CACHE_DEPTH))) as u8;

            let node = self.read_depth(self.v4_offset, CACHE_DEPTH, 0, &b);
            self.ipv4_bits_cache.push(node);
        }
    }

    fn is_ipv4(&self) -> bool {
        self.meta.ip_version & 1 == 1
    }

    fn is_ipv6(&self) -> bool {
        self.meta.ip_version & 2 == 2
    }

    pub fn find(&self, addr :&[u8], lang:&str, _cache:bool) -> Result<IpInfo> {
        let node = self.find0(addr)?;
        // if cache && self.result_cache.len() > 0 {
        //     let x = self.result_cache.get(&node);
        //     match x {
        //         Some(i) => {
        //             return Ok(*i);
        //         }
        //         None => {}
        //     }
        // }
        let v = self.resolve_node(node, lang)?;
        Ok(Self::build_info(&v, &self.meta.fields))
    }

    fn build_info(val:&Vec<String>, fields: &Vec<String>) -> IpInfo {
        let mut info = IpInfo{
            country: "".to_string(),
            district: "".to_string(),
            province: "".to_string(),
            city: "".to_string(),
            asn: "".to_string(),
            isp: "".to_string(),
            continent: "".to_string(),
            line: "".to_string(),
        };
        for i in 0..fields.len() {
            match fields[i].as_str(){
                "country_name"       => info.country = val[i].to_string(),
                "region_name" => info.province = val[i].to_string(),
                "city_name" => info.city = val[i].to_string(),
                "isp_domain"=> info.isp = val[i].to_string(),
                "asn" => info.asn = val[i].to_string(),
                "line" => info.line = val[i].to_string(),
                "district" => info.district = val[i].to_string(),
                "continent_code" => {
                    match val[i].as_str() {
                        "AS" => info.continent = "亚洲".to_string(),
                        "EU" => info.continent = "欧洲".to_string(),
                        "NA" => info.continent = "北美洲".to_string(),
                        "SA" => info.continent = "南美洲".to_string(),
                        "AF" => info.continent = "非洲".to_string(),
                        "OC" => info.continent = "大洋洲".to_string(),
                        "AN" => info.continent = "南极洲".to_string(),
                        _ => {}
                    }
                },
                _ => {}
            };
        }
        info
    }

    fn find0(&self, ip: &[u8]) -> Result<i32> {
        if ip.len() == 16 {
            if !self.is_ipv6() {
                return Err(Error::new(ErrorKind::InvalidData, "no support ipv6"));
            }
        } else if ip.len() == 4 {
            if !self.is_ipv4() {
                return Err(Error::new(ErrorKind::InvalidData, "no support ipv4"));
            }
        } else {
            return Err(Error::new(ErrorKind::InvalidData, "invalid ip"));
        }
        self.find_node(ip)
    }

    fn resolve_node(&self, node: i32, lang:&str) -> Result<Vec<String>> {
        let off = self.meta.languages.get(lang);
        match off {
            None => return Err(Error::new(ErrorKind::InvalidData, "invalid language")),
            Some(_) => {
                let resolved = node - self.node_count + self.node_count * 8;
                if resolved >= self.file_size {
                    return Err(Error::new(ErrorKind::InvalidData, "database resolve error"));
                }

                let resolved: usize = resolved as usize;

                let size = BigEndian::read_u16(&self.data[resolved ..resolved + 2]) as usize;

                if self.data.len() < resolved + 2 + size {
                    return Err(Error::new(ErrorKind::InvalidData, "database resolve error"));
                }
                let str = String::from_utf8_lossy(&self.data[resolved  + 2..resolved  + 2 + size ]);
                let a = str.split("\t");
                let r = a.map(|i| i.to_string()).collect();
                Ok(r)
            }
        }
    }

    fn find_node(&self, b: &[u8]) -> Result<i32> {
        let mut node = 0;
        let bit = b.len() * 8;
        if bit == 32 {
            node = self.v4_offset;
        }
        let mut i = 0;
        if self.ipv4_bits_cache.len()>0 && bit == 32 {
            let index = Self::bytes_to_index(b);
            node = self.ipv4_bits_cache[index as usize];
            if node >= self.node_count {
                return Ok(node);
            }
            i = CACHE_DEPTH;
        }
        node = self.read_depth(node, bit as i32, i, b);
        if node >= self.node_count {
            return Ok(node);
        }

        return Err(io::ErrorKind::NotFound.into());
    }

    fn bytes_to_index(b: &[u8]) -> i32 {
        let i1 = ((0xFF & (b[0] as i32)) << 8) >> (16 - CACHE_DEPTH);
        let i2 = (0xFF & (b[1] as i32)) >> (16 - CACHE_DEPTH);
        i1 | i2
    }

    fn read_depth(&self, node0: i32, depth:i32, start:i32, b:&[u8]) -> i32 {
        let mut node = node0;
        for i in start..depth {
            if node >= self.node_count {
                break;
            }
            let index = 1 & ((0xFF & b[(i / 8) as usize] as usize) >> 7 - (i % 8));
            node = self.read_node(node, index as i32);
        }
        node
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn meta_deser() {

    }
}
